netcore/System.Private.CoreLib/src/System/Environment.Unix.cs

```diff
@@ -13,36 +13,91 @@ namespace System
 {
 	partial class Environment
 	{
+		private static Dictionary<string, string> s_environment;
+
 		static string GetEnvironmentVariableCore (string variable)
 		{
-			using (var h = RuntimeMarshal.MarshalString (variable)) {
-				return internalGetEnvironmentVariable_native (h.Value);
+			Debug.Assert(variable != null);
+
+			if (s_environment == null) {
+				using (var h = RuntimeMarshal.MarshalString (variable)) {
+					return internalGetEnvironmentVariable_native (h.Value);
+				}
+			}
+
+			lock (s_environment) {
+				variable = TrimStringOnFirstZero (variable);
```

Comment:
TrimStringOnFirstZero should not be under lock
Should it be called in managed for internalGetEnvironmentVariable_native too?

Reply:
The trimming on zero merely mimics what the C APIs already do implicitly in native code. They accept zero-terminated strings.